Basic token swapping

KieranWoodTombola opened the pull request:
-Feature Token Swapping
-Feature On Swapping tokens, use token locations to check for matches and update the board

Alpha fade tween acts as a placeholder to highlight tokens that have successfully been updated. Rapid Token swapping will break the highlight

KieranWoodTombola committed:
Feature - Swap Tokens on the grid

KieranWoodTombola committed:
Grid match logic bufixing

KieranWoodTombola committed:
Responding to comments

KieranWoodTombola committed:
Merge remote-tracking branch 'origin/main' into basic_token_swapping

AndrewMollett commented on src/scripts/core/models/column.ts:
```diff
@@ -37,6 +37,17 @@ export class Column extends Container{
+    }
+
+    public replaceAllTokens(newTokens: Token[]): void {
+        if(newTokens.length != this.tokens.length) {
```
You need to use !== here, not !=. What you have here is a 'truthy/falsy' check, which is less strict on what it counts as true/false, using !== makes it a much more strict check

AndrewMollett commented on src/scripts/core/models/grid.ts:
```diff
@@ -36,41 +32,73 @@ export class Grid extends Container {
+        if((this.activeToken === 0) && (this.firstTokenLocation[0] === -1) && (this.firstTokenLocation[0] === -1)) {
+            targetToken.highLight();
+            this.activeToken = 1;
+            this.firstTokenLocation = targetToken.getLocation();
```
I wonder, do you need to base this on the token location? By identifying the token within the array, you should be able to store which tokens are highlighted right?

Which also means your calculations below may be simplified

liamf1986 commented on src/scripts/core/models/column.ts:
```diff
@@ -37,6 +37,17 @@ export class Column extends Container{
+
+    public replaceAllTokens(newTokens: Token[]): void {
+        if(newTokens.length != this.tokens.length) {
+            console.log("error on replaceAllTokens: input length does not match");
+            return;
```
Looks like this is reporting an error. console.log is fine for basic reporting while developing but there are a couple alternatives you may not yet know about.

- `console.error()` - basically the same as log, but it shows as red in the console to highlight the issue better.
- `throw new Error('message')` or just `throw 'message'` - the technically correct way to handle errors as it allows you to detect and handle your errors with a `try {} catch {}` block like you've probably seen in other languages.

liamf1986 commented on src/scripts/core/models/grid.ts:
```diff
@@ -8,22 +8,18 @@ export class Grid extends Container {
     private columns: Column[];
     private activeToken: number;
-
+    private firstTokenLocation: number[];
```
Is this intended to always represent an x and y co-ordinate in your grid?

If so it could be typed better than just `number[]` which can be an array of any length.

- You could make it a tuple (`private location: [number, number]`).
This is still an array and will work with what you currently have, except TypeScript will prevent you from accidentally assigning/reading anything other than the first 2.

- You could also use the type `IPointData` which Pixi provides and is just a basic `{ x: number; y: number }` structure.

liamf1986 commented on src/scripts/core/models/grid.ts:
```diff
@@ -8,22 +8,18 @@ export class Grid extends Container {
         this.columns = [];
         this.activeToken = 0;
+        const firstTokenLocation = [-1, -1];
+        this.firstTokenLocation = firstTokenLocation;
```
You don't appear to use the `const firstTokenLocation` other than immediately assigning it to `this.firstTokenLocation`. Looks like you could remove that line and just set `this.firstTokenLocation = [-1, -1];`

liamf1986 commented on src/scripts/core/models/grid.ts:
```diff
@@ -36,41 +32,73 @@ export class Grid extends Container {
+
+        //on SecondClick
+        if(this.activeToken === 1) {
+            targetToken.highLight();
+            this.activeToken = 2;
+
+            //Identify first Token
+            const firstX = this.firstTokenLocation[0];
+            const firstY = this.firstTokenLocation[1];
+            const firstTokenCopy = this.getToken(firstX, firstY);
+            const firstSkIndex = firstTokenCopy.getSkIndex();
+
+            //Identify second Token
+            const secondX = targetToken.getLocation()[0];
+            const secondY = targetToken.getLocation()[1];
+            const secondSkIndex = this.getToken(secondX, secondY).getSkIndex();
 
-    private swapPrepare(): void {
+            // //Swap Tokens
+            const tempFirstToken = this.getToken(firstX, firstY);
+            const tempSecondToken = this.getToken(secondX, secondY);
+            
+
+            //Swapping isn't working as expected
+            let targetFirst = this.getToken(firstX, firstY);
+            targetFirst.setToken(secondSkIndex);
+            let targetSecond = this.getToken(secondX, secondY);
+            targetSecond.setToken(firstSkIndex);
+
+            this.matchCheck(this.getToken(firstX, firstY));
+            this.matchCheck(this.getToken(secondX, secondY));
+
+            this.nukeBoard();
+
+            this.firstTokenLocation = [-1, -1];
+            this.activeToken = 0;
+            return;
+        }
```
Definitely some cleanup that can be done in this whole section.

For example you call `this.getToken(firstX, firstY)` 4 times in here and assign that same result into 3 different variables.

I think Mollets point above stands true.
You are receiving the clicked Token in this event so it may be cleaner to just:
- on first click, store the first token in a variable.
- on second click you already have the second token passed through so can read the first token from your variable and never need to call `getToken` at all.
- just make a copy of each tokens skin, then re-assign each token with the other skin.

liamf1986 commented on src/scripts/core/models/grid.ts:
```diff
@@ -145,26 +175,29 @@ export class Grid extends Container {
+            const matched = targetColumn.getAllTokens().filter(isMatched);
             matched.forEach(token => {
                 token.shuffleSkin();
+                token.highLight();
+                newTokenArray.push(token);
             });
-            const newColumn = column.getAllTokens();
-            matched.forEach(token => {
-                newColumn.push(token);
-            });
+
+            const unmatched = targetColumn.getAllTokens().filter(isNotMatched);
             unmatched.forEach(token => {
-                newColumn.push(token);
+                newTokenArray.push(token);
             });
```
I'm not sure exactly how this is working but it looks like your columns are constantly getting bigger each time this is called.

Since you create newTokenArray, by getting all token in the column:
`const newTokenArray = targetColumn.getAllTokens();`

Then you appear to be pushing both the matched and unmatched tokens into the array again. Nothing is ever being removed from it so it's essentially doubling in size each time?

liamf1986 replied:
I should also point out that you have an issue with referencing here.

Javascript passes objects as references and arrays are objects.
What that means is in `targetColumn.getAllToken()` where it just returns `this.tokens` (an array), it's returning a reference to that array not a copy of it.

So in the above code everytime you change `newTokenArray` you are also changing the underlying array inside the original column. Which is why you are not seeing an error inside your `replaceAllTokens(newTokens: Token[]): void` error check, because you are passing your reference back into it and effectively comparing the same array to itself.

KieranWoodTombola committed:
Refactor Grid model

-Stronger typing for Class objects.
    As per PR feedback, adopted Touples to simplify
    Token fetching and manipulation

-Refactored the grid update procedures.
    Previously, Grid would search for matches using
    onClick's provided Token
    The Previous method fails to account for matches set on init
    or chain matches following user interaction

    Now, resolveMatches() declares the rows and columns to search.
    matchLine() marks combinations of Tokens for later removal.

    Token Removal (or at least the animations showing removal/
    replacement) can then be done at Column level

KieranWoodTombola committed:
Fix scene

- changes grid size set in scene from 5 to 6
    A size of 6 allows testing for grid positions with multiple
    valid combinations of tokens with different skins
    on the same row/column

KieranWoodTombola committed:
Refactor Column

-set tokens to public for easier access
    removed get method for tokens

KieranWoodTombola committed:
Refactor Token

-small changes to improve readability

-removed simple get methods and set some properties to public

-Removed unnecessary comments and obsolete features

KieranWoodTombola commented:
Refactored Models to allow Grid to better handle Token searching as per previous comments

tombolabrad commented on src/scripts/core/models/grid.ts:
```diff
@@ -6,24 +6,16 @@ import { Token } from "./token";
-
+    private columns: Column[] = [];
+    private tokensSelected: number = 0;
+    private selectedTokens: [Token | undefined, Token | undefined] = [undefined, undefined];
```
could this not just be a Token[] which you change the length of?
this would mean not having to have undefined as part of the typing

to get tokensSelected you could instead call selectedTokens.length

KieranWoodTombola committed:
Refactor Grid

- Removed tokensSelected as a tracker for tokens clicked by users
    based actions off of selectedTokens defined/undefined

KieranWoodTombola committed:
Refactor Column

- Small type changes and comment removal

KieranWoodTombola committed:
Refactor Column

- Small type changes and comment removal

KieranWoodTombola committed:
Refactor Grid

-Fixed matchLine and refactored grid processing so that
    rows and columns are properly transformed prior to token
    movement and management
